### PR TITLE
fix: end span if outgoing http request ends prematurely

### LINKED ISSUE #1564: Aborted outgoing HTTP requests are not recorded as spans
Currently, the agent listens for the `end` event to be emitted by the request object in order to end the span, but if `req.abort()` is called, this event is obviously never emitted.

### PR DESCRIPTION
**Pull request by watson**
Fixes #1564

### Checklist

- [x] Implement code
- [x] Add tests
- [x] Commit message follows [commit guidelines](https://github.com/elastic/apm-agent-nodejs/blob/master/CONTRIBUTING.md#commit-message-guidelines)

### CODE REVIEW

**Comment by watson** [comment]
jenkins, run the tests please

File: lib/instrumentation/http-shared.js [hunk 1]
```diff
@@ -182,13 +182,28 @@ exports.traceOutgoingRequest = function (agent, moduleName, method) {
...
         agent.logger.warn('unable to identify http.ClientRequest url %o', { id: id })
       })
 
+      // In case the request is ended prematurely
+      endOfStream(req, function () {
+        if (span.ended) return
+
+        agent.logger.debug('unexpected end of outgoing http request %o', { id: id })
+
+        span.setHttpContext({
```
Comment (vigneshshanmugam) [inline lib/instrumentation/http-shared.js:198]: The tests fail as the destination context info is not added to these spans

Comment (vigneshshanmugam) [reply lib/instrumentation/http-shared.js:207]: Fixed it properly now.

File: lib/instrumentation/http-shared.js [hunk 2]
```diff
@@ -182,33 +182,35 @@ exports.traceOutgoingRequest = function (agent, moduleName, method) {
...
-        ins.bindEmitter(res)
-
-        res.prependListener('end', onEnd)
+      // In case the request is ended prematurely
+      endOfStream(req, function () {
+        // End span only when request is aborted as EOS fires for
+        // 'abort', 'close', 'end' events which might end span prematurely
+        if (req.aborted) {
+          if (span.ended) return
+          agent.logger.debug('unexpected end of outgoing http request %o', { id: id })
```
Comment (axw) [inline lib/instrumentation/http-shared.js:198]: why is this unexpected? isn't it expected that the request stream is ended when the request is aborted...?

Comment (vigneshshanmugam) [reply lib/instrumentation/http-shared.js:198]: My guess its not common to abort requests in the middle of the stream and could also be due to socket reset by peer. I will rename it like this `intercepted http.IncomingMessage abort event` to keep it in sync.

File: lib/instrumentation/http-shared.js [hunk 3]
```diff
@@ -191,24 +192,22 @@ exports.traceOutgoingRequest = function (agent, moduleName, method) {
...
-      function onresponse (res) {
-        // Work around async_hooks bug in Node.js 12.0 - 12.2 (https://github.com/nodejs/node/pull/27477)
-        ins._recoverTransaction(span.transaction)
-
-        agent.logger.debug('intercepted http.ClientRequest response event %o', { id: id })
-        ins.bindEmitter(res)
-
-        res.prependListener('end', onEnd)
+      // In case the request is ended prematurely
+      function onAbort (type) {
```
Comment (vigneshshanmugam) [inline lib/instrumentation/http-shared.js:196]: @axw I changed the original implementation which relied on end of steam package and used our previous emit logic for handling the `abort` events, its easier and does not add event listeners for unnecessary events.

**Comment by vigneshshanmugam** [comment]
Windows test failure is unrelated.

**Pull request merged**